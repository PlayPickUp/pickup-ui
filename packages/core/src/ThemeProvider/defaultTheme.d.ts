import { DefaultTheme } from "../../../../types";
export declare const defaultTheme: DefaultTheme;
//# sourceMappingURL=defaultTheme.d.ts.map
