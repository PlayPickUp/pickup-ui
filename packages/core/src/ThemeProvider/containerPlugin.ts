/*
 *  Adds container params to all selectors
 *  Example: 'h1' becomes 'body #PickUpUI h1'
 */

import { Plugin } from "jss";

export function containUI(withReset: boolean, container?: string): Plugin {
  const prefix = container ? container : "body #PickUpUI ";

  // 'rule' is left untyped because the Rule import from jss is not recognizing all child properties
  function onProcessRule(rule): Plugin["onProcessRule"] {
    const parent = rule.options.parent;
    if (
      rule.type !== "style" ||
      rule.selectorText.includes(prefix) ||
      (parent && parent.type === "keyframes") ||
      (withReset && parent && parent.key === "@global")
    )
      return;

    if (rule.selectorText.includes(",")) {
      const allSelectors = rule.selectorText.split(",");
      const transformedSelectors = allSelectors.map((selector: string) =>
        selector.includes(prefix) ? selector : prefix + selector
      );
      rule.selectorText = transformedSelectors.join(",");
    } else {
      rule.selectorText = prefix + rule.selectorText;
    }
  }
  return { onProcessRule };
}
