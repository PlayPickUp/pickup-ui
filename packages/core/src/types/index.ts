import { FieldInputProps, FormikErrors, FormikTouched } from "formik";
import React, {
  ChangeEvent,
  CSSProperties,
  MouseEvent,
  ReactElement,
  ReactNode,
  SetStateAction,
  Dispatch,
} from "react";

/**
 *
 * Theme Specific Types
 *
 */
export interface ThemeProviderProps {
  children: any;
  withReset?: boolean;
  theme?: DefaultTheme;
}

export interface PurpleMap {
  light: string;
  base: string;
  dark: string;
}

export interface GreenMap {
  light: string;
  base: string;
}

export interface GreyMap {
  light: string;
  base: string;
  dark: string;
}

export interface RedMap {
  light: string;
  base: string;
}

export interface DefaultThemeColors {
  primary: PurpleMap;
  secondary: GreyMap;
  green: GreenMap;
  purple: PurpleMap;
  black: string;
  white: string;
  success: string;
  grey: GreyMap;
  red: RedMap;
  error: string;
}

export interface FontFamilies {
  body: string;
  headline: string;
}

interface FontProperties {
  fontFamily: string;
  fontWeight: number | string;
  fontStyle?: string;
  fontSize: number | string;
  lineHeight: string;
  letterSpacing: string;
  textTransform?: string;
  color: string;
}

export interface FontStyles {
  title: FontProperties;
  heading2: FontProperties;
  heading3: FontProperties;
  body: FontProperties;
  body2: FontProperties;
}

export interface ResponsiveFonts {
  desktop: FontStyles;
  mobile: FontStyles;
}

export interface DefaultTypography {
  fontFamilies: FontFamilies;
  fontStyles: ResponsiveFonts;
}
export type CreateShadow = (
  a: number,
  b: number,
  c: number,
  d: number,
  e: number,
  f: number,
  g: number,
  h: number,
  i: number,
  j: number,
  k: number,
  l: number
) => string;

export type DefaultSpacing = {
  base: number;
};

export type DefaultShadows = Array<string>;

export interface DefaultZIndex {
  overlay: number;
}

export interface DefaultGradients {
  main: string;
  hover: string;
}

export type MediaQuery = (minWidth: number) => string;

export interface Breakpoints {
  small: number;
  medium: number;
  large: number;
}

export type DefaultTheme = {
  borderRadius: number;
  breakpoints: Breakpoints;
  mediaQuery: MediaQuery;
  colors: DefaultThemeColors;
  gradients: DefaultGradients;
  shadows: DefaultShadows;
  spacing: DefaultSpacing;
  typography: DefaultTypography;
  zIndex: DefaultZIndex;
};

/**
 *
 * Button Component
 *
 */

export type ButtonVariant = "fit" | "full";

export interface ButtonProps {
  children?: ReactNode;
  disabled?: boolean;
  style?: React.CSSProperties;
  variant?: ButtonVariant;
  onClick?: (e?: React.MouseEvent) => unknown;
  element?: keyof JSX.IntrinsicElements;
  to?: string;
  href?: string;
  className?: string;
  useSubmit?: boolean;
  submitText?: string;
  color?: "primary" | "secondary" | "light";
}

/**
 *
 * Typography Component
 *
 */

export interface TypographyProps {
  children: ReactNode;
  variant?:
    | "title"
    | "body"
    | "body2"
    | "heading2"
    | "heading3"
    | "heading4"
    | "heading5"
    | "heading6"
    | "span";
  className?: string;
  element?: "p" | "span" | "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
  style?: CSSProperties;
  color?: string;
  useUnescape?: boolean;
}

export type TypographyElementMap = Record<string, keyof JSX.IntrinsicElements>;

/**
 *
 * Fab (Floating Action Button) Component
 *
 */

export interface FabProps {
  className?: string;
  innerClassName?: string;
  innerStyle?: React.CSSProperties;
  style?: React.CSSProperties;
  icon: any;
  title: string;
  size?: "small" | "large";
  onClick: (e: MouseEvent) => unknown;
  disabled?: boolean;
  disablePopOver?: boolean;
  color?: string;
  iconColor?: string;
}

/**
 *
 * Table Component
 *
 */
export interface HeadCell {
  disablePadding: boolean;
  id: keyof TableData;
  label: string;
  numeric?: boolean;
  width?: number | undefined;
}

export type TableData = Record<any, any>;

export type TableOrder = "asc" | "desc";
export interface DataTableProps {
  headCells: HeadCell[];
  rows: Array<Record<string | number, any>>;
  defaultSortColumn: string;
  tableTitle: string;
  actionToolbar: (props: any) => ReactElement;
}

export interface EnhancedTableProps {
  classes: ReturnType<any>;
  numSelected: number;
  onRequestSort: (event: React.MouseEvent<unknown>, property: any) => void;
  onSelectAllClick: (event: React.ChangeEvent<HTMLInputElement>) => void;
  order: TableOrder;
  orderBy: string;
  rowCount: number;
  headCells: HeadCell[];
}

export interface EnhancedTableToolbarProps {
  numSelected: number;
  tableTitle: string;
  actionToolbar: (props: any) => ReactElement;
  selected: string[];
  setSelected: Dispatch<SetStateAction<string[]>>;
}

/**
 *
 * Icon Component
 *
 */
export interface IconBaseProps {
  children: ReactElement;
  color?: string;
  size?: number;
  className?: string;
  style?: React.CSSProperties;
}

export interface BaseFormikFields {
  id: string;
  name: string;
  field: FieldInputProps<any>;
  form: Record<any, any>;
}

export type BaseFormikHandleChange = (e: ChangeEvent) => unknown;

export interface BaseFormikFieldsWithHandleChange extends BaseFormikFields {
  handleChange: BaseFormikHandleChange;
}

/**
 *
 * TextInput Component
 *
 */
export interface TextInputProps extends BaseFormikFieldsWithHandleChange {
  className?: string;
  innerClassName?: string;
  innerStyle?: React.CSSProperties;
  style?: React.CSSProperties;
  placeholder?: string;
  label?: string;
}

/**
 *
 * Label Component
 *
 */
export interface LabelProps {
  htmlFor: string;
  children: string | ReactNode;
  className?: string;
  style?: React.CSSProperties;
  innerClassName?: string;
  innerStyle?: React.CSSProperties;
}

/**
 *
 * TextArea Component
 *
 */
export interface TextAreaProps extends BaseFormikFieldsWithHandleChange {
  className?: string;
  style?: React.CSSProperties;
  innerClassName?: string;
  innerStyle?: React.CSSProperties;
  label?: string;
}

/**
 *
 * Select Component
 *
 */ export interface SelectItem {
  label: string;
  value: string;
}

export type BaseFormikSetFieldValue = (
  field: string,
  value: any,
  shouldValidate?: boolean | undefined
) => unknown;

export interface SelectProps extends BaseFormikFields {
  className?: string;
  items: SelectItem[];
  label: string;
  placeholder?: string;
  style?: React.CSSProperties;
}

export interface DropdownStyleProps {
  isOpen: boolean;
}

/**
 *
 * NestedInput Component
 *
 */
export interface NestedInputProps extends BaseFormikFieldsWithHandleChange {
  buttonText: string;
  placeholder: string;
}

/**
 *
 * MultiSelect Component
 *
 */
export interface MultiSelectProps extends BaseFormikFields {
  items: SelectItem[];
  label: string;
}

/**
 *
 * Individual Icon Component
 *
 */
export interface IconProps {
  color?: string;
  className?: string;
  style?: React.CSSProperties;
}

/**
 *
 * Chip Component
 *
 */
export interface ChipProps {
  disabled?: boolean;
  onClick?: (e: React.MouseEvent) => unknown;
  color?: "default" | "primary" | "secondary";
  className?: string;
  style?: React.CSSProperties;
  isActive?: boolean;
  label: string;
  element?: keyof JSX.IntrinsicElements;
  href?: string;
}

/**
 *
 * Paper Component
 *
 */
export interface PaperProps {
  padding?: number;
  children: unknown;
  style?: React.CSSProperties;
  className?: string;
}

/**
 *
 * FormError Component
 *
 */
export interface FormErrorProps {
  errors: FormikErrors<unknown>;
  touched: FormikTouched<unknown>;
  name: string;
}

/**
 *
 * Loader Component
 *
 */
export interface LoaderProps {
  className?: string;
  style?: React.CSSProperties;
  delay?: number;
  text?: string;
}

/**
 *
 * Picker Button
 *
 */
export interface PickerButtonProps {
  className?: string;
  displayText: string;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
  style?: React.CSSProperties;
  showResult?: boolean;
  result: number | null;
  isPick: boolean;
}
