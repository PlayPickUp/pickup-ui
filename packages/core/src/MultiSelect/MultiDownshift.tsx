import React, { ReactNode } from "react";
import Downshift, {
  DownshiftState,
  ControllerStateAndHelpers,
} from "downshift";
import without from "lodash/without";

import { SelectItem } from "../types";
import { FieldInputProps } from "formik";

export interface MultiDownshiftProps {
  onChange: (field: string, value: string) => void;
  itemToString: (item: SelectItem) => string;
  field: FieldInputProps<any>;
  render?: any;
}

export interface MultiDownshiftState {
  selectedItems: SelectItem[];
}

class MultiDownshift extends React.Component<
  MultiDownshiftProps,
  MultiDownshiftState
> {
  state: MultiDownshiftState = { selectedItems: [] };

  stateReducer = (
    state: DownshiftState<SelectItem>,
    changes: any | Record<any, boolean | string | number>
  ): Record<any, boolean | string | number> => {
    switch (changes.type) {
      case Downshift.stateChangeTypes.keyDownEnter:
      case Downshift.stateChangeTypes.clickItem:
        return {
          ...changes,
          highlightedIndex: state.highlightedIndex,
          isOpen: true,
          inputValue: "",
        };
      default:
        return changes;
    }
  };

  handleSelection = (
    selectedItem: SelectItem,
    downshift: ControllerStateAndHelpers<SelectItem>
  ): void => {
    const callOnChange = () => {
      const { onChange, field } = this.props;
      const { selectedItems } = this.state;
      if (onChange) {
        const transformedItems = selectedItems.map(
          (item: SelectItem) => item.value
        );
        onChange(field.name, transformedItems.join(","));
        this.getStateAndHelpers(downshift);
      }
    };
    if (this.state.selectedItems.includes(selectedItem)) {
      this.removeItem(selectedItem, callOnChange);
    } else {
      this.addSelectedItem(selectedItem, callOnChange);
    }
  };

  removeItem = (item: SelectItem, cb?: () => unknown): void => {
    const { onChange, field } = this.props;

    const fieldArray = without(field.value.split(","), item.value);
    onChange(field.name, fieldArray.toString());

    this.setState(({ selectedItems }) => {
      return {
        selectedItems: selectedItems.filter((i) => i !== item),
      };
    }, cb);
  };

  addSelectedItem = (item: SelectItem, cb: () => unknown): void => {
    this.setState(
      ({ selectedItems }) => ({
        selectedItems: [...selectedItems, item],
      }),
      cb
    );
  };

  // @ts-ignore
  getRemoveButtonProps = ({ onClick, item, ...props } = {}): Record<
    any,
    unknown
  > => {
    return {
      onClick: (e) => {
        onClick && onClick(e);
        e.stopPropagation();
        this.removeItem(item);
      },
      ...props,
    };
  };

  getStateAndHelpers = (
    downshift: ControllerStateAndHelpers<SelectItem>
  ): Record<any, unknown> => {
    const { selectedItems } = this.state;
    const { getRemoveButtonProps, removeItem } = this;
    return {
      getRemoveButtonProps,
      removeItem,
      selectedItems,
      ...downshift,
    };
  };

  render(): ReactNode {
    const { render, children = render, ...props } = this.props;
    return (
      <Downshift
        {...props}
        stateReducer={this.stateReducer}
        onChange={this.handleSelection}
        selectedItem={null}
      >
        {(downshift) => children(this.getStateAndHelpers(downshift))}
      </Downshift>
    );
  }
}

export default MultiDownshift;
