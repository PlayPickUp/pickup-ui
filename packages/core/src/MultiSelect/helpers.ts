import { matchSorter } from "match-sorter";
import { SelectItem } from "../types";

export const getItems = (
	allItems: SelectItem[],
	filter: string
): SelectItem[] => {
	return filter
		? matchSorter(allItems, filter.toLowerCase(), { keys: ["value"] })
		: allItems;
};
