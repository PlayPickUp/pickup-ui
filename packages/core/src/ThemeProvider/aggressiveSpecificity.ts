/* This is based on the unsupported jss-increase-specificity plugin at https://github.com/iamstarkov/jss-increase-specificity
JSS's original createGenerateId has been adapted here to use logic from that plugin. /*

// From the original plugin:
/**
 * use :not(#\20), :not(.\20) and :not(\20) instead of generating unlikely
 * appearing ids…
 * — twitter.com/subzey/status/829050478721896448
 * Rationale: \20 is a css escape for U+0020 Space, and neither classname,
 * nor id, nor tagname can contain a space
 * — twitter.com/subzey/status/829051085885153280
 */

import { Rule, CreateGenerateId } from "jss";
import { JssStyleSheet } from "../types";

var selector = ":not(#\\20)";
const repeat = 3;
const maxRules = 1e10;

export const increaseSpecificity: CreateGenerateId = (options = {}) => {
  let ruleCounter = 0;
  console.log({ options });
  return (rule: Rule, sheet?: JssStyleSheet): string => {
    ruleCounter += 1;
    if (ruleCounter > maxRules) {
      console.warn(
        false,
        `[JSS] You might have a memory leak. Rule counter is at ${ruleCounter}.`
      );
    }
    const jssId = String(sheet.options.jss.id) ?? "";
    const prefix = sheet.options.classNamePrefix ?? "";

    let suffix = "";
    for (let i = 0; i < repeat; i++) {
      suffix += selector;
    }

    return `${prefix + rule.key}${
      jssId ? `-${jssId}` : ""
    }-${ruleCounter}${suffix}`;
  };
};
