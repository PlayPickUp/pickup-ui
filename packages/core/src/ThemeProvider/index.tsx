import React from "react";
import { jss, JssProvider, ThemeProvider as JSSThemeProvider } from "react-jss";
import { containUI } from "./containerPlugin";
import merge from "lodash/merge";

import { defaultTheme } from "./defaultTheme";

import { ThemeProviderProps } from "../types";
import GlobalsAndReset from "./GlobalsAndReset";

function generateRandom(digitCount) {
  return Math.floor(
      Math.pow(10, digitCount - 1) +
      Math.random() * 9 * Math.pow(10, digitCount - 1)
  );
}

const ThemeProvider: React.FC<ThemeProviderProps> = ({
  children,
  withReset = true,
  theme,
  ...rest
}) => {
  jss.use(containUI(withReset));
  const mergedTheme = theme ? merge(defaultTheme, theme) : defaultTheme;

  const random = generateRandom(5)

  return (
    <div id="PickUpUI">
      <JssProvider jss={jss} classNamePrefix={`PU${random}--`}>
        <JSSThemeProvider theme={mergedTheme} {...rest}>
          <GlobalsAndReset>{children}</GlobalsAndReset>
        </JSSThemeProvider>
      </JssProvider>
    </div>
  );
};

export default ThemeProvider;
