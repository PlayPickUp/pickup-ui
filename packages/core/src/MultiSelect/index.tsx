import React, { KeyboardEvent, useRef } from "react";
import MultiDownshift from "./MultiDownshift";
import classNames from "classnames";
import { createUseStyles, useTheme } from "react-jss";
import ChevronDown from "../Icon/icons//ChevronDown";
import SubtractCircle from "../Icon/icons/SubtractCircle";

import Icon from "../Icon";
import { getItems } from "./helpers";
import { DefaultTheme, MultiSelectProps, SelectItem } from "../types";
import Label from "../Label";
import FormError from "../FormError";

const useStyles = createUseStyles((theme: DefaultTheme) => ({
  root: {
    position: "relative",
  },
  tagWrapper: {
    display: "flex",
    justifyContent: "flex-start",
    alignItems: "center",
    flexFlow: "row wrap",
    position: "relative",
    width: "100%",
    height: "auto",
  },
  tag: {
    display: "flex",
    flexFlow: "row nowrap",
    justifyContent: "space-between",
    alignItems: "center",
    position: "relative",
    width: "auto",
    backgroundColor: theme.colors.purple.light,
    border: `1px solid ${theme.colors.purple.dark}`,
    borderRadius: "55px",
    padding: `${theme.spacing.base}px ${theme.spacing.base * 3.5}px`,
    margin: `${theme.spacing.base * 2}px ${theme.spacing.base}px 0`,
    fontSize: 12,
    fontFamily: theme.typography.fontFamilies.body,
    color: theme.colors.purple.dark,
    textAlign: "center",
    cursor: "default",
    "& span": {
      display: "block",
      position: "relative",
      lineHeight: 1,
    },
  },
  inputWrapper: {
    position: "relative",
    height: "auto",
    width: "100%",
    fontFamily: theme.typography.fontFamilies.body,
    fontSize: 15,
    lineHeight: "18px",
    letterSpacing: "0.1px",
    color: theme.colors.black,
    backgroundColor: theme.colors.white,
    border: `1px solid ${theme.colors.grey.base}`,
    borderRadius: theme.borderRadius,
    padding: `0 ${theme.spacing.base * 5}px 0 ${theme.spacing.base * 4}px`,
    textAlign: "left",
    "&:focus, &:active": {
      border: `1px solid ${theme.colors.grey.base}`,
      outline: "none",
    },
    boxSizing: "border-box",
  },
  input: {
    position: "relative",
    width: "100%",
    height: 44,
    border: "none",
    fontFamily: theme.typography.fontFamilies.body,
    fontSize: 15,
    lineHeight: "18px",
    letterSpacing: "0.1px",
    color: theme.colors.black,
    backgroundColor: theme.colors.white,
    "&:active, &:focus": {
      outline: "none",
    },
  },
  iconButton: {
    display: "inline-block",
    "-webkit-appearance": "none",
    "-moz-appearance": "none",
    userSelect: "none",
    textDecoration: "none",
    fontFamily: "inherit",
    fontSize: "inherit",
    color: "inherit",
    cursor: "pointer",
    border: "none",
    backgroundColor: "transparent",
    borderColor: "transparent",
    padding: "0",
    margin: "0",
    "&:disabled": {
      cursor: "not-allowed",
    },
    position: "absolute",
    right: `${theme.spacing.base * 3}px`,
    top: `${theme.spacing.base * 3.25}px`,
    "&:active, &:focus": {
      outline: "none",
    },
  },
  icon: {
    position: "relative",
    transition: "all 250ms ease-out",
    transform: "rotate(0deg)",
  },
  iconOpen: {
    transform: "rotate(180deg)",
  },
  dropdown: {
    position: "relative",
    top: -3,
    height: "200px",
    overflow: "scroll",
    width: "100%",
    fontFamily: theme.typography.fontFamilies.body,
    fontSize: 14,
    lineHeight: "18px",
    letterSpacing: "0.1px",
    color: theme.colors.black,
    backgroundColor: theme.colors.white,
    border: `1px solid ${theme.colors.grey.base}`,
    borderTop: "transparent",
    borderRadius: theme.borderRadius,
    borderTopLeftRadius: 0,
    borderTopRightRadius: 0,
    textAlign: "left",
    "&:focus, &:active": {
      border: `1px solid ${theme.colors.grey.base}`,
      borderTop: "transparent",
      outline: "none",
    },
  },
  dropdownItem: {
    padding: `${theme.spacing.base * 4}px`,
  },
  subtractIcon: {
    marginLeft: theme.spacing.base,
    right: -7,
  },
}));

const MultiSelect: React.FC<MultiSelectProps> = ({
  items,
  label,
  ...props
}) => {
  const classes = useStyles();
  const theme: DefaultTheme = useTheme();
  const input = useRef(null);

  return (
    <>
      <MultiDownshift
        onChange={props.form.setFieldValue}
        itemToString={(item: SelectItem) => (item ? item.label : "")}
        field={props.field}
      >
        {({
          getInputProps,
          getToggleButtonProps,
          getMenuProps,
          // note that the getRemoveButtonProps prop getter and the removeItem
          // action are coming from MultiDownshift composibility for the win!
          getRemoveButtonProps,
          removeItem,
          isOpen,
          inputValue,
          selectedItems,
          getItemProps,
          highlightedIndex,
        }) => (
          <div className={classes.root}>
            <Label htmlFor={props.field.name}>{label}</Label>
            <div style={{ position: "relative" }}>
              <div
                className={classes.inputWrapper}
                style={{ position: "relative" }}
              >
                {selectedItems.length > 0 ? (
                  <div className={classes.tagWrapper}>
                    {selectedItems.map((item: SelectItem, i: number) => (
                      <div key={`${item.value}${i}`} className={classes.tag}>
                        <span>{item.label}</span>
                        <button
                          {...getRemoveButtonProps({
                            item,
                          })}
                        >
                          <Icon
                            className={classes.subtractIcon}
                            size={12}
                            color={theme.colors.purple.dark}
                          >
                            <SubtractCircle />
                          </Icon>
                        </button>
                      </div>
                    ))}
                  </div>
                ) : null}
                <input
                  {...getInputProps({
                    ref: input,
                    placeholder: "Select items...",
                    className: classes.input,
                    onKeyDown(event: KeyboardEvent) {
                      if (
                        event.key === "Backspace" &&
                        !inputValue &&
                        selectedItems.length > 0
                      ) {
                        removeItem(selectedItems[selectedItems.length - 1]);
                      }
                    },
                    ...props.field,
                  })}
                />
                <button
                  {...getToggleButtonProps({
                    className: classes.iconButton,
                  })}
                >
                  <Icon
                    className={classNames({
                      [classes.icon]: true,
                      [classes.iconOpen]: isOpen,
                    })}
                    color={theme.colors.grey.base}
                  >
                    <ChevronDown />
                  </Icon>
                </button>
              </div>
            </div>
            <ul
              {...getMenuProps({
                className: classes.dropdown,
                style: { display: isOpen ? "block" : "none" },
              })}
            >
              <div>
                <input
                  {...getInputProps({
                    placeholder: "Search items...",
                    id: "search",
                    className: classes.input,
                  })}
                />
                {isOpen
                  ? getItems(items, inputValue).map(
                      (item: SelectItem, index: number) => (
                        <li
                          key={`${item.value}${index}`}
                          {...getItemProps({
                            item,
                            index,
                            className: classes.dropdownItem,
                            style: {
                              backgroundColor:
                                highlightedIndex === index
                                  ? theme.colors.purple.light
                                  : theme.colors.white,
                              fontWeight: selectedItems.includes(item)
                                ? "bold"
                                : "normal",
                            },
                          })}
                        >
                          {item.label}
                        </li>
                      )
                    )
                  : null}
              </div>
            </ul>
          </div>
        )}
      </MultiDownshift>
      <FormError
        errors={props.form.errors}
        touched={props.form.touched}
        name={props.field.name}
      />
    </>
  );
};

export default MultiSelect;
