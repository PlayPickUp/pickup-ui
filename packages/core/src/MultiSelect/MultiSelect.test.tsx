/**
 *
 * NOTE: These tests will fail in MultiDownshift without the plugin
 * '@babel/plugin-proposal-class-properties in the babel.config file
 *
 **/

import React from "react";
import { render, fireEvent, cleanup, waitFor } from "@testing-library/react";
import { Formik, Field, Form } from "formik";

import ThemeProvider from "../ThemeProvider";
import MultiSelect from ".";
import { SelectItem } from "../types";

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// Props
const items: SelectItem[] = [
  {
    label: "Option 1",
    value: "option1",
  },
  {
    label: "Option 2",
    value: "option2",
  },
  {
    label: "Option 3",
    value: "option3",
  },
];

const FormComponent: React.FC<{
  onSubmit: jest.Mock;
}> = ({ onSubmit }) => {
  const handleSubmit = async (values) => {
    await sleep(500);
    onSubmit(values);
  };
  return (
    <ThemeProvider>
      <div>
        <Formik
          initialValues={{
            selectMany: "",
          }}
          onSubmit={handleSubmit}
        >
          <Form>
            <Field
              id="selectMany"
              name="selectMany"
              label="MultiSelect Test"
              items={items}
              component={MultiSelect}
            />
            <button type="submit">Submit</button>
          </Form>
        </Formik>
      </div>
    </ThemeProvider>
  );
};

const handleSubmit = jest.fn();
// We could break these out into a testHelpers file, if they're reused elsewhere
const downArrow = { key: "Down", code: "Down", charCode: 40 };
const enterKey = { key: "Enter", code: "Enter", charCode: 13 };

test("MultiSelect renders without crashing, matches snapshot", () => {
  const { container } = render(<FormComponent onSubmit={handleSubmit} />);
  expect(container).toMatchSnapshot();
});

test("Props are passed and rendered", () => {
  const { getByText, getByPlaceholderText } = render(
    <FormComponent onSubmit={handleSubmit} />
  );

  expect(getByText("MultiSelect Test")).toBeTruthy();

  const input = getByPlaceholderText("Search & Select items...");
  // Down arrow opens menu to reveal options
  fireEvent.keyDown(input, downArrow);
  expect(getByText("Option 2")).toBeTruthy();
  cleanup;
});

test("Selected item gets added to list", () => {
  const { container, getByPlaceholderText } = render(
    <FormComponent onSubmit={handleSubmit} />
  );

  const input = getByPlaceholderText("Search & Select items...");
  fireEvent.keyDown(input, downArrow);
  fireEvent.keyDown(input, enterKey);

  expect(container.getElementsByTagName("span").length).toEqual(1);
  cleanup;
});

test("Values are passed to form submission", async () => {
  const { getByPlaceholderText, getByRole } = render(
    <FormComponent onSubmit={handleSubmit} />
  );

  const input = getByPlaceholderText("Search & Select items...");
  fireEvent.keyDown(input, downArrow);
  fireEvent.keyDown(input, enterKey);

  fireEvent.click(getByRole("button", { name: /submit/i }));

  await waitFor(() =>
    expect(handleSubmit).toHaveBeenCalledWith({
      selectMany: "option1",
    })
  );
});
