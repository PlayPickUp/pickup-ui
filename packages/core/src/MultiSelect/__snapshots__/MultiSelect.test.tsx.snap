// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect renders without crashing, matches snapshot 1`] = `
<div>
  <div
    id="PickUpUI"
  >
    <div>
      <form
        action="#"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="downshift-0-label"
          class="PU--root-0-2-1"
          role="combobox"
        >
          <div
            data-testid="label-wrapper"
          >
            <label
              class="PU--label-0-2-12"
              data-testid="label"
              for="selectMany"
            >
              MultiSelect Test
            </label>
          </div>
          <div
            style="position: relative;"
          >
            <div
              class="PU--inputWrapper-0-2-4"
              style="position: relative;"
            >
              <input
                aria-autocomplete="list"
                aria-labelledby="downshift-0-label"
                autocomplete="off"
                class="PU--input-0-2-5"
                id="downshift-0-input"
                placeholder="Search & Select items..."
                value=""
              />
              <button
                aria-haspopup="true"
                aria-label="open menu"
                class="PU--iconButton-0-2-6"
                data-toggle="true"
                role="button"
                type="button"
              >
                <div
                  class="PU--root-0-2-13 PU--root-d0-0-2-14 PU--icon-0-2-7"
                  data-testid="icon"
                >
                  <svg
                    height="140"
                    version="1.1"
                    viewBox="0 0 140 140"
                    width="140"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlns:xlink="http://www.w3.org/1999/xlink"
                  >
                    <g
                      transform="matrix(5.833333333333333,0,0,5.833333333333333,0,0)"
                    >
                      <path
                        d="M23.25,7.311,12.53,18.03a.749.749,0,0,1-1.06,0L.75,7.311"
                        fill="none"
                        stroke="#9D9D9E"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="1.5"
                      />
                    </g>
                  </svg>
                </div>
              </button>
            </div>
          </div>
          <ul
            aria-labelledby="downshift-0-label"
            class="PU--dropdown-0-2-9"
            id="downshift-0-menu"
            role="listbox"
            style="display: none;"
          >
            <div />
          </ul>
        </div>
        <button
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;
